// rsuite/lib/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _isUndefined = require('lodash/isUndefined');

var _isUndefined2 = _interopRequireDefault(_isUndefined);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  title: _propTypes2.default.string,
  inline: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  checked: _propTypes2.default.bool,
  defaultChecked: _propTypes2.default.bool,
  onChange: _propTypes2.default.func,
  inputRef: _propTypes2.default.func,
  value: _propTypes2.default.any // eslint-disable-line react/forbid-prop-types
};

var Checkbox = function (_React$Component) {
  _inherits(Checkbox, _React$Component);

  function Checkbox(props) {
    _classCallCheck(this, Checkbox);

    var _this = _possibleConstructorReturn(this, (Checkbox.__proto__ || Object.getPrototypeOf(Checkbox)).call(this, props));

    _this.state = {
      checked: props.checked || props.defaultChecked
    };
    _this.handleChange = _this.handleChange.bind(_this);
    return _this;
  }

  _createClass(Checkbox, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.checked !== this.props.checked) {
        this.setState({
          checked: nextProps.checked
        });
      }
    }
  }, {
    key: 'handleChange',
    value: function handleChange(event) {
      var _props = this.props,
          onChange = _props.onChange,
          disabled = _props.disabled,
          value = _props.value;

      var target = event.target;
      var checked = !this.state.checked;

      if (disabled) {
        return;
      }

      this.setState({ checked: checked }, function () {
        var nextValue = (0, _isUndefined2.default)(value) ? checked : target.value;
        onChange && onChange(nextValue, event);
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          inline = _props2.inline,
          disabled = _props2.disabled,
          className = _props2.className,
          onChange = _props2.onChange,
          children = _props2.children,
          title = _props2.title,
          inputRef = _props2.inputRef,
          style = _props2.style,
          checked = _props2.checked,
          props = _objectWithoutProperties(_props2, ['inline', 'disabled', 'className', 'onChange', 'children', 'title', 'inputRef', 'style', 'checked']);

      var nextChecked = (0, _isUndefined2.default)(checked) ? this.state.checked : checked;
      var classes = (0, _classnames2.default)('checkbox', {
        'checkbox-inline': inline
      }, className);

      var checkboxClasses = (0, _classnames2.default)('checker', {
        disabled: disabled
      });

      var input = _react2.default.createElement(
        'span',
        { className: (0, _classnames2.default)('checkbox-wrapper', { checked: nextChecked }) },
        _react2.default.createElement('input', _extends({}, props, {
          type: 'checkbox',
          ref: inputRef,
          disabled: disabled,
          onChange: this.handleChange
        }))
      );

      return _react2.default.createElement(
        'div',
        {
          className: classes,
          style: style
        },
        _react2.default.createElement(
          'div',
          {
            className: checkboxClasses,
            role: 'button'
          },
          _react2.default.createElement(
            'label',
            { title: title },
            input,
            children
          )
        )
      );
    }
  }]);

  return Checkbox;
}(_react2.default.Component);

Checkbox.displayName = 'Checkbox';
Checkbox.propTypes = propTypes;

exports.default = Checkbox;
