// rsuite/lib/DropdownMenuItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _elementType = require('rsuite-utils/lib/propTypes/elementType');

var _elementType2 = _interopRequireDefault(_elementType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  divider: _propTypes2.default.bool,
  active: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  onSelect: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  eventKey: _propTypes2.default.any, // eslint-disable-line react/forbid-prop-types
  componentClass: _elementType2.default
};

var defaultProps = {
  componentClass: 'a'
};

var DropdownMenuItem = function (_React$Component) {
  _inherits(DropdownMenuItem, _React$Component);

  function DropdownMenuItem(props) {
    _classCallCheck(this, DropdownMenuItem);

    var _this = _possibleConstructorReturn(this, (DropdownMenuItem.__proto__ || Object.getPrototypeOf(DropdownMenuItem)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  _createClass(DropdownMenuItem, [{
    key: 'handleClick',
    value: function handleClick(event) {
      var _props = this.props,
          onSelect = _props.onSelect,
          eventKey = _props.eventKey,
          disabled = _props.disabled,
          onClick = _props.onClick;

      if (disabled) {
        event.preventDefault();
        return;
      }

      onSelect && onSelect(eventKey, this.props, event);
      onClick && onClick(event);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          children = _props2.children,
          divider = _props2.divider,
          onSelect = _props2.onSelect,
          active = _props2.active,
          disabled = _props2.disabled,
          className = _props2.className,
          eventKey = _props2.eventKey,
          Component = _props2.componentClass,
          props = _objectWithoutProperties(_props2, ['children', 'divider', 'onSelect', 'active', 'disabled', 'className', 'eventKey', 'componentClass']);

      var classes = (0, _classnames2.default)({
        active: active,
        disabled: disabled
      }, className);

      if (divider) {
        return _react2.default.createElement('li', { role: 'separator', className: 'divider' });
      }

      return _react2.default.createElement(
        'li',
        _extends({}, props, {
          role: 'presentation',
          className: classes
        }),
        _react2.default.createElement(
          Component,
          {
            tabIndex: '-1',
            onClick: this.handleClick
          },
          children
        )
      );
    }
  }]);

  return DropdownMenuItem;
}(_react2.default.Component);

DropdownMenuItem.displayName = 'DropdownMenuItem';
DropdownMenuItem.propTypes = propTypes;
DropdownMenuItem.defaultProps = defaultProps;

exports.default = DropdownMenuItem;
