// rsuite/lib/DropdownToggle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _omit = require('lodash/omit');

var _omit2 = _interopRequireDefault(_omit);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _SafeAnchor = require('./SafeAnchor');

var _SafeAnchor2 = _interopRequireDefault(_SafeAnchor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  noCaret: _propTypes2.default.bool,
  title: _propTypes2.default.node,
  useAnchor: _propTypes2.default.bool
};

var DorpdownToggle = function (_React$Component) {
  _inherits(DorpdownToggle, _React$Component);

  function DorpdownToggle() {
    _classCallCheck(this, DorpdownToggle);

    return _possibleConstructorReturn(this, (DorpdownToggle.__proto__ || Object.getPrototypeOf(DorpdownToggle)).apply(this, arguments));
  }

  _createClass(DorpdownToggle, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          noCaret = _props.noCaret,
          useAnchor = _props.useAnchor,
          title = _props.title,
          className = _props.className,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['noCaret', 'useAnchor', 'title', 'className', 'children']);

      var caret = noCaret ? null : _react2.default.createElement('span', { className: 'caret' });
      var Component = useAnchor ? _SafeAnchor2.default : _Button2.default;
      var elementProps = useAnchor ? (0, _omit2.default)(props, 'block') : props;
      return _react2.default.createElement(
        Component,
        _extends({}, elementProps, {
          className: (0, _classnames2.default)('dropdown-toggle', className),
          role: 'button'
        }),
        title || children,
        caret
      );
    }
  }]);

  return DorpdownToggle;
}(_react2.default.Component);

DorpdownToggle.propTypes = propTypes;

exports.default = DorpdownToggle;
