// rsuite/lib/DropdownMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _isEqual = require('lodash/isEqual');

var _isEqual2 = _interopRequireDefault(_isEqual);

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ReactChildren = require('./utils/ReactChildren');

var _ReactChildren2 = _interopRequireDefault(_ReactChildren);

var _isNullOrUndefined = require('./utils/isNullOrUndefined');

var _isNullOrUndefined2 = _interopRequireDefault(_isNullOrUndefined);

var _createChainedFunction = require('./utils/createChainedFunction');

var _createChainedFunction2 = _interopRequireDefault(_createChainedFunction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  prefixClass: _propTypes2.default.string,
  pullRight: _propTypes2.default.bool,
  onSelect: _propTypes2.default.func,
  activeKey: _propTypes2.default.any // eslint-disable-line react/forbid-prop-types
};

var defaultProps = {
  prefixClass: 'dropdown',
  pullRight: false,
  onSelect: null,
  activeKey: null
};

var DorpdownMenu = function (_React$Component) {
  _inherits(DorpdownMenu, _React$Component);

  function DorpdownMenu() {
    _classCallCheck(this, DorpdownMenu);

    return _possibleConstructorReturn(this, (DorpdownMenu.__proto__ || Object.getPrototypeOf(DorpdownMenu)).apply(this, arguments));
  }

  _createClass(DorpdownMenu, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          pullRight = _props.pullRight,
          children = _props.children,
          className = _props.className,
          activeKey = _props.activeKey,
          onSelect = _props.onSelect,
          prefixClass = _props.prefixClass,
          props = _objectWithoutProperties(_props, ['pullRight', 'children', 'className', 'activeKey', 'onSelect', 'prefixClass']);

      var classes = (0, _classnames2.default)(_defineProperty({}, prefixClass + '-menu-right', pullRight), prefixClass + '-menu', className);

      var items = _ReactChildren2.default.mapCloneElement(children, function (item) {
        var _item$props = item.props,
            eventKey = _item$props.eventKey,
            active = _item$props.active,
            onItemSelect = _item$props.onSelect;

        var displayName = (0, _get2.default)(item, ['type', 'displayName']);
        if (displayName === 'DropdownMenuItem' || displayName === 'NavItem') {
          return {
            onSelect: (0, _createChainedFunction2.default)(onSelect, onItemSelect),
            active: (0, _isNullOrUndefined2.default)(activeKey) ? active : (0, _isEqual2.default)(activeKey, eventKey)
          };
        }
        return null;
      });

      return _react2.default.createElement(
        'ul',
        _extends({}, props, {
          className: classes,
          role: 'menu'
        }),
        items
      );
    }
  }]);

  return DorpdownMenu;
}(_react2.default.Component);

DorpdownMenu.propTypes = propTypes;
DorpdownMenu.defaultProps = defaultProps;

exports.default = DorpdownMenu;
