// rsuite/lib/ModalDialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _decorate = require('./utils/decorate');

var _decorate2 = _interopRequireDefault(_decorate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  prefixClass: _propTypes2.default.string,
  dialogClassName: _propTypes2.default.string,
  dialogStyle: _propTypes2.default.object // eslint-disable-line react/forbid-prop-types
};

var defaultProps = {
  prefixClass: 'modal'
};

var ModalDialog = function (_React$Component) {
  _inherits(ModalDialog, _React$Component);

  function ModalDialog() {
    _classCallCheck(this, ModalDialog);

    return _possibleConstructorReturn(this, (ModalDialog.__proto__ || Object.getPrototypeOf(ModalDialog)).apply(this, arguments));
  }

  _createClass(ModalDialog, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          style = _props.style,
          children = _props.children,
          dialogClassName = _props.dialogClassName,
          dialogStyle = _props.dialogStyle,
          prefixClass = _props.prefixClass,
          props = _objectWithoutProperties(_props, ['className', 'style', 'children', 'dialogClassName', 'dialogStyle', 'prefixClass']);

      var modalStyle = _extends({
        display: 'block'
      }, style);

      var modalClasses = (0, _classnames2.default)(prefixClass, className);
      var dialogClasses = (0, _classnames2.default)(_extends({}, (0, _decorate.getClassNames)(this.props), _defineProperty({}, prefixClass, false)), prefixClass + '-dialog', dialogClassName);

      return _react2.default.createElement(
        'div',
        _extends({
          title: null,
          role: 'dialog',
          style: modalStyle,
          className: modalClasses
        }, props),
        _react2.default.createElement(
          'div',
          {
            className: dialogClasses,
            style: dialogStyle
          },
          _react2.default.createElement(
            'div',
            { className: prefixClass + '-content', role: 'document' },
            children
          )
        )
      );
    }
  }]);

  return ModalDialog;
}(_react2.default.Component);

ModalDialog.propTypes = propTypes;
ModalDialog.defaultProps = defaultProps;

exports.default = (0, _decorate2.default)({
  size: true
})(ModalDialog);
