// rsuite/lib/Nav.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _isEqual = require('lodash/isEqual');

var _isEqual2 = _interopRequireDefault(_isEqual);

var _NavItem = require('./NavItem');

var _NavItem2 = _interopRequireDefault(_NavItem);

var _NavDropdown = require('./NavDropdown');

var _NavDropdown2 = _interopRequireDefault(_NavDropdown);

var _ReactChildren = require('./utils/ReactChildren');

var _ReactChildren2 = _interopRequireDefault(_ReactChildren);

var _isNullOrUndefined = require('./utils/isNullOrUndefined');

var _isNullOrUndefined2 = _interopRequireDefault(_isNullOrUndefined);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  prefixClass: _propTypes2.default.string,
  tabs: _propTypes2.default.bool,
  pills: _propTypes2.default.bool,
  justified: _propTypes2.default.bool,
  stacked: _propTypes2.default.bool,
  pullRight: _propTypes2.default.bool,
  activeKey: _propTypes2.default.any, // eslint-disable-line react/forbid-prop-types
  onSelect: _propTypes2.default.func
};

var defaultProps = {
  prefixClass: 'nav',
  tabs: false,
  pills: false,
  justified: false,
  stacked: false,
  pullRight: false,
  activeKey: undefined,
  onSelect: undefined
};

var contextTypes = {
  navbar: _propTypes2.default.bool
};

var Nav = function (_React$Component) {
  _inherits(Nav, _React$Component);

  function Nav() {
    _classCallCheck(this, Nav);

    return _possibleConstructorReturn(this, (Nav.__proto__ || Object.getPrototypeOf(Nav)).apply(this, arguments));
  }

  _createClass(Nav, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          prefixClass = _props.prefixClass,
          tabs = _props.tabs,
          pills = _props.pills,
          stacked = _props.stacked,
          justified = _props.justified,
          pullRight = _props.pullRight,
          className = _props.className,
          children = _props.children,
          onSelect = _props.onSelect,
          activeKey = _props.activeKey,
          props = _objectWithoutProperties(_props, ['prefixClass', 'tabs', 'pills', 'stacked', 'justified', 'pullRight', 'className', 'children', 'onSelect', 'activeKey']);

      var classes = (0, _classnames2.default)(prefixClass, (_classNames = {}, _defineProperty(_classNames, prefixClass + 'bar-right', pullRight), _defineProperty(_classNames, prefixClass + 'bar-nav', this.context.navbar), _defineProperty(_classNames, prefixClass + '-pills', pills), _defineProperty(_classNames, prefixClass + '-tabs', tabs), _defineProperty(_classNames, prefixClass + '-stacked', stacked), _defineProperty(_classNames, prefixClass + '-justified', justified), _classNames), className);

      var items = _ReactChildren2.default.mapCloneElement(children, function (item) {
        var _item$props = item.props,
            eventKey = _item$props.eventKey,
            active = _item$props.active;

        if (item.type.displayName !== 'NavItem') {
          return null;
        }
        return {
          onSelect: onSelect,
          active: (0, _isNullOrUndefined2.default)(activeKey) ? active : (0, _isEqual2.default)(activeKey, eventKey)
        };
      });

      return _react2.default.createElement(
        'ul',
        _extends({}, props, { className: classes }),
        items
      );
    }
  }]);

  return Nav;
}(_react2.default.Component);

Nav.propTypes = propTypes;
Nav.defaultProps = defaultProps;
Nav.contextTypes = contextTypes;

Nav.Item = _NavItem2.default;
Nav.Dropdown = _NavDropdown2.default;

exports.default = Nav;
