// rsuite/lib/NavItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _elementType = require('rsuite-utils/lib/propTypes/elementType');

var _elementType2 = _interopRequireDefault(_elementType);

var _createChainedFunction = require('./utils/createChainedFunction');

var _createChainedFunction2 = _interopRequireDefault(_createChainedFunction);

var _SafeAnchor = require('./SafeAnchor');

var _SafeAnchor2 = _interopRequireDefault(_SafeAnchor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  active: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  onClick: _propTypes2.default.func,
  onSelect: _propTypes2.default.func,
  eventKey: _propTypes2.default.any, // eslint-disable-line react/forbid-prop-types
  componentClass: _elementType2.default
};

var defaultProps = {
  href: '',
  active: false,
  disabled: false,
  componentClass: _SafeAnchor2.default,
  onClick: undefined,
  onSelect: undefined,
  eventKey: undefined

};

var NavItem = function (_React$Component) {
  _inherits(NavItem, _React$Component);

  function NavItem(props) {
    _classCallCheck(this, NavItem);

    var _this = _possibleConstructorReturn(this, (NavItem.__proto__ || Object.getPrototypeOf(NavItem)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  _createClass(NavItem, [{
    key: 'handleClick',
    value: function handleClick(event) {
      var _props = this.props,
          onSelect = _props.onSelect,
          disabled = _props.disabled,
          eventKey = _props.eventKey;

      if (onSelect && !disabled) {
        onSelect(eventKey, this.props, event);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          active = _props2.active,
          disabled = _props2.disabled,
          onClick = _props2.onClick,
          className = _props2.className,
          style = _props2.style,
          eventKey = _props2.eventKey,
          children = _props2.children,
          Component = _props2.componentClass,
          props = _objectWithoutProperties(_props2, ['active', 'disabled', 'onClick', 'className', 'style', 'eventKey', 'children', 'componentClass']);

      var classes = (0, _classnames2.default)({
        active: active,
        disabled: disabled
      }, className);

      return _react2.default.createElement(
        'li',
        {
          role: 'presentation',
          className: classes,
          style: style
        },
        _react2.default.createElement(
          Component,
          _extends({}, props, {
            disabled: disabled,
            onClick: (0, _createChainedFunction2.default)(onClick, this.handleClick)
          }),
          children
        )
      );
    }
  }]);

  return NavItem;
}(_react2.default.Component);

NavItem.propTypes = propTypes;
NavItem.defaultProps = defaultProps;
NavItem.displayName = 'NavItem';

exports.default = NavItem;
