// rsuite/lib/PaginationButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _elementType = require('rsuite-utils/lib/propTypes/elementType');

var _elementType2 = _interopRequireDefault(_elementType);

var _createChainedFunction = require('./utils/createChainedFunction');

var _createChainedFunction2 = _interopRequireDefault(_createChainedFunction);

var _SafeAnchor = require('./SafeAnchor');

var _SafeAnchor2 = _interopRequireDefault(_SafeAnchor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  eventKey: _propTypes2.default.any, // eslint-disable-line react/forbid-prop-types
  onSelect: _propTypes2.default.func,
  disabled: _propTypes2.default.bool,
  active: _propTypes2.default.bool,
  onClick: _propTypes2.default.func,
  componentClass: _elementType2.default
};

var defaultProps = {
  eventKey: undefined,
  onSelect: undefined,
  onClick: undefined,
  active: false,
  disabled: false,
  componentClass: _SafeAnchor2.default
};

var PaginationButton = function (_React$Component) {
  _inherits(PaginationButton, _React$Component);

  function PaginationButton(props) {
    _classCallCheck(this, PaginationButton);

    var _this = _possibleConstructorReturn(this, (PaginationButton.__proto__ || Object.getPrototypeOf(PaginationButton)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    return _this;
  }

  _createClass(PaginationButton, [{
    key: 'handleClick',
    value: function handleClick(event) {
      var _props = this.props,
          disabled = _props.disabled,
          onSelect = _props.onSelect,
          eventKey = _props.eventKey;

      if (disabled) {
        return;
      }

      onSelect && onSelect(eventKey, event);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          active = _props2.active,
          disabled = _props2.disabled,
          onClick = _props2.onClick,
          Component = _props2.componentClass,
          className = _props2.className,
          style = _props2.style,
          onSelect = _props2.onSelect,
          eventKey = _props2.eventKey,
          props = _objectWithoutProperties(_props2, ['active', 'disabled', 'onClick', 'componentClass', 'className', 'style', 'onSelect', 'eventKey']);

      return _react2.default.createElement(
        'li',
        {
          className: (0, _classnames2.default)(className, { active: active, disabled: disabled }),
          style: style
        },
        _react2.default.createElement(Component, _extends({}, props, {
          disabled: disabled,
          onClick: (0, _createChainedFunction2.default)(onClick, this.handleClick)
        }))
      );
    }
  }]);

  return PaginationButton;
}(_react2.default.Component);

PaginationButton.propTypes = propTypes;
PaginationButton.defaultProps = defaultProps;

exports.default = PaginationButton;
